Add docs for `data-turbo-method` and `data-turbo-confirm` for `link_to`

julianfssen opened the pull request:
<!--
Thanks for contributing to Rails!

Please do not make *Draft* pull requests, as they still send
notifications to everyone watching the Rails repo.

Create a pull request when it is ready for review and feedback
from the Rails team :).

If your pull request affects documentation or any non-code
changes, guidelines for those changes are [available
here](https://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#contributing-to-the-rails-documentation)

About this template

The following template aims to help contributors write a good description for their pull requests.
We'd like you to provide a description of the changes in your pull request (i.e. bugs fixed or features added), motivation behind the changes, and complete the checklist below before opening a pull request.

Feel free to discard it if you need to (e.g. when you just fix a typo). -->

### Motivation / Background

Rails 7 ships with Turbo enabled by default. Instead of using `data-method` and `data-confirm`, Turbo now uses `data-turbo-method` and `data-turbo-confirm` to perform a link visit with the specified HTTP method and show confirmation dialog respectively.

The deprecated legacy options are documented **but the new options are not.**

### Detail

This commit documents the `data-turbo-method` and `data-turbo-confirm` for the `link_to` method. The `button_to` documentation has also been updated to reference the new `link_to` options.

### Additional information

There is an argument that this change is unrelated to Rails, since it is a [Turbo-specific behavior](https://turbo.hotwired.dev/handbook/drive#performing-visits-with-a-different-method). 

However, I feel it needs to be documented because:

1) Turbo is enabled by default so people who were used to the legacy `data-method` and `data-confirm` may not be aware of it when upgrading or starting a new Rails 7 project.
2) The new options are referenced in the [getting started guide](https://guides.rubyonrails.org/getting_started.html#deleting-an-article), so it makes sense to reference it in the API documentation as well.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

julianfssen committed:
Add docs for `data-turbo-method` and `data-turbo-confirm` for `link_to`

Rails 7 ships with Turbo enabled by default. Instead of using
`data-method` and `data-confirm`, Turbo now uses `data-turbo-method` and
`data-turbo-confirm` to perform a link visit with the specified HTTP
method and show confirmation dialog respective.

The deprecated legacy options are documented but the new options are
not.

This commit documents the `data-turbo-method` and `data-turbo-confirm`
for the `link_to` method.

The `button_to` documentation has also been updated to reference the
new `link_to` options.

ghiculescu commented:
These examples are helpful, but I think it would also be good to link to the Turbo Handbook.

ghiculescu commented on actionview/lib/action_view/helpers/url_helper.rb:
```diff
@@ -172,6 +172,25 @@ def _filtered_referrer # :nodoc:
+      #   link_to "Delete profile", @profile, data: { turbo_method: :delete }
+      #   # => <a href="/profiles/1" data-turbo-method="delete">Delete profile</a>
+      #
+      # Since forms cannot appear inside another form, +button_to+ <i>cannot</i> be used here as it returns a form.
```
Hmm. Doesn't `data-turbo-method` also generate a form at runtime?

julianfssen replied:
Great question, @ghiculescu ! 

`data-turbo-method` _does_ generate a form at runtime. But, [it appends the generated form as a child of `document.body`](https://github.com/hotwired/turbo/blob/f8b0e17a0c605080de0097112a5cf408abef9bb1/src/observers/form_link_click_observer.ts#L63) instead of it returning it directly like `button_to`, which is a neat shortcut to make non-`GET` requests work in forms.

[The Turbo handbook does mention this](https://turbo.hotwired.dev/handbook/drive#performing-visits-with-a-different-method):

> The link will get converted into a hidden form next to the a element in the DOM. This means that the link can’t appear inside another form, as you can’t have nested forms.

I'm not entirely sure if that's accurate since this type of link works in my testing. The source code also suggests that the form is appended to `document.body` rather than the link where `data-turbo-method` is called.

ghiculescu replied:
> I'm not entirely sure if that's accurate since this type of link works in my testing. The source code also suggests that the form is appended to document.body rather than the link where data-turbo-method is called.

Should probalby update the Turbo handbook.

julianfssen replied:
Good idea, @ghiculescu ! I will create a PR shortly for updating the Turbo handbook regarding `data-turbo-method`.

julianfssen replied:
Opened in: https://github.com/hotwired/turbo-site/pull/142

ghiculescu commented on actionview/lib/action_view/helpers/url_helper.rb:
```diff
@@ -172,6 +172,25 @@ def _filtered_referrer # :nodoc:
+      # A common use case is to insert links to perform update or delete operations (i.e. non-<tt>GET</tt> requests) in a form.
+      # For example, you might use +link_to+ in an edit form to delete a record:
+      #
+      #   link_to "Delete profile", @profile, data: { turbo_method: :delete }
```
This example should probably be a `button_to`. Is there a more appropriate example we can use here?

julianfssen replied:
You're right, @ghiculescu ! I've added an expanded example of `button_to` that fails, and a short explanation on why `link_to` with `data-turbo-method` should be used instead in this commit: https://github.com/rails/rails/pull/47505/commits/1e568cd47103fd7ce8aa1f3fbe16779a9e73e23f

julianfssen commented:
@ghiculescu , thank you for taking the time to review this PR! I've addressed your comments and looking forward to hearing back again when you're free. Thanks again!

p8 commented:
Hi @julianfssen 👋 .
Could you wrap the lines at 100 characters?

julianfssen commented:
> Hi @julianfssen 👋 . Could you wrap the lines at 100 characters?

Thank you for feedback, @p8 ! 

I've added a commit (https://github.com/rails/rails/pull/47505/commits/666c29a8e4493223ce60bf5a5b7d05e6add390f2) to wrap the docs at 100 characters, which I believe starts from the comment character `#`? Do correct me if I'm wrong on this.

My bad as I wasn't aware of this guideline previously. I appreciate the feedback!

julianfssen commented:
PS: I've committed several `fixup` commits for easier referencing in comments, but will be squashing them once (and if) this PR is approved. Thank you everybody!

skipkayhil commented on actionview/lib/action_view/helpers/url_helper.rb:
```diff
@@ -172,6 +172,39 @@ def _filtered_referrer # :nodoc:
+      #   button_to "Delete profile", @profile, method: :delete
+      #
+      # Since forms cannot appear inside another form,
+      # +button_to+ will not work here as it returns a form.
```
Yeah I like that idea a lot

ghiculescu replied:
I will make a PR

zzak commented on actionview/lib/action_view/helpers/url_helper.rb:
```diff
@@ -172,6 +172,39 @@ def _filtered_referrer # :nodoc:
+      # The <tt>data-turbo-confirm</tt> HTML attribute adds a confirmation dialog to the link:
+      #
+      #   link_to "Visit Other Site", "https://rubyonrails.org/", data: { turbo_confirm: "Are you sure?" }
+      #   # => <a href="https://rubyonrails.org/" data-turbo-confirm="Are you sure?">
+      #          Visit Other Site
+      #        </a>
```
```suggestion
      #   # => <a href="https://rubyonrails.org/" data-turbo-confirm="Are you sure?">Visit Other Site</a>
```

No need to break up the lines here IMO

zzak commented on actionview/lib/action_view/helpers/url_helper.rb:
```diff
@@ -239,6 +269,10 @@ def link_to(name = nil, options = nil, html_options = nil, &block)
       # HTTP verb via the +:method+ option within +html_options+.
       #
+      # If the HTML button generated from +button_to+ does not work with your layout, you can
+      # consider using the +link_to+ method with the <tt>data-turbo-method</tt>
```
```suggestion
      # consider using the +link_to+ method with the +data-turbo-method+
```

RDoc deals with this fine using `+` for fixed-width:

```
$ echo "+kebab-case+" | rdoc --pipe
# => <p><code>kebab-case</code></p>
```

zzak commented on actionview/lib/action_view/helpers/url_helper.rb:
```diff
@@ -172,6 +172,39 @@ def _filtered_referrer # :nodoc:
+      # ==== Turbo
+      #
+      # Rails 7 ships with Turbo enabled by default. The <tt>data-turbo-method</tt> HTML attribute
+      # can be used to perform a Turbo link visit with a different HTTP method. {Consult the
+      # Turbo Handbook for more information on performing visits with different
+      # methods.}[https://turbo.hotwired.dev/handbook/drive#performing-visits-with-a-different-method]
```
```suggestion
      # can be used to perform a Turbo link visit with a different HTTP method.
      # {Consult the Turbo Handbook for more information on performing visits with different
      # methods.}[https://turbo.hotwired.dev/handbook/drive#performing-visits-with-a-different-method]
```

julianfssen commented:
Thank you for the feedback, @zzak ! I've addressed your comments and also squashed the commits for this PR. Appreciate the feedback again!

brunoprietog commented:
It sounds good to add it to the documentation, but this is really bad accessibility practice. Links should only be used for get actions. Anything else should be done with a button, for which there is the button_to helper with which you can achieve the same result by changing the method and action. If you need to execute actions inside forms, you can change them with a button containing the formmethod and formaction attributes without using links for it.

Considering this, I don't know if it is good to keep expanding the use of this by showing it in Rails documentation and guides.

What do you think? I know that this is not well known by many people, that is why I am not in favor of continuing to encourage its use.

julianfssen commented:
> It sounds good to add it to the documentation, but this is really bad accessibility practice. Links should only be used for get actions. Anything else should be done with a button, for which there is the button_to helper with which you can achieve the same result by changing the method and action. If you need to execute actions inside forms, you can change them with a button containing the formmethod and formaction attributes without using links for it.
> 
> Considering this, I don't know if it is good to keep expanding the use of this by showing it in Rails documentation and guides.
> 
> What do you think? I know that this is not well known by many people, that is why I am not in favor of continuing to encourage its use.

Thanks for the constructive feedback, @brunoprietog ! I agree this is bad for accessibility which is also why I added a note about it in the PR.

That said, using links to perform non-`GET` actions has been a thing ever since `ujs` ([which is referenced in the API docs as well](https://github.com/rails/rails/blob/7c70791470fc517deb7c640bead9f1b47efb5539/actionview/lib/action_view/helpers/url_helper.rb#L173)), so I don't think it it's new knowledge for a lot of Rails users.

The reason why I created this PR was because I was caught out by `data-method` and `data-confirm` not working when I upgraded my Rails app. I checked the Rails docs for `link_to` but could only find the deprecation note. I then discovered the solution through several Github issues and the Turbo handbook, which made me think it's a good idea to mention in the docs.

Of course, the right thing to do as you mentioned (which I agree) is to use a `button` or restructure the link to not use `data-turbo-method`. However, I think it'd be more convenient for many people to change `data-method` to `data-turbo-method` instead of rewriting their forms and links. 

The PR is also not adding new behavior or suggesting the usage of a private API; `data-turbo-method` is exposed for this purpose and [is mentioned in the Turbo handbook](https://turbo.hotwired.dev/handbook/drive#performing-visits-with-a-different-method), which makes me feel it's fine to mention it in the Rails docs as well.

Let me know what you think of this! Thank you again for the feedback!

julianfssen commented:
@ghiculescu @zzak thank you for the feedback! My bad for re-bumping this PR. May I know if the changes made above are good now? Thank you!

julianfssen commented on actionview/lib/action_view/helpers/url_helper.rb:
```diff
@@ -172,6 +172,37 @@ def _filtered_referrer # :nodoc:
+      #   # => <a href="/profiles/1" data-turbo-method="delete">Delete profile</a>
+      #
+      # Using <tt>data-turbo-method</tt> is not ideal for accessibility
+      # but it is useful in scenarios like the above.
```
That makes sense, @zzak ! I initially added this excerpt in as it was mentioned in the [Turbo handbook](https://turbo.hotwired.dev/handbook/drive#performing-visits-with-a-different-method). But, I agree it's unnecessary in the Rails docs.

julianfssen commented on actionview/lib/action_view/helpers/url_helper.rb:
```diff
@@ -239,6 +267,10 @@ def link_to(name = nil, options = nil, html_options = nil, &block)
       #
+      # If the HTML button generated from +button_to+ does not work with your layout, you can
+      # consider using the +link_to+ method with the +data-turbo-method+
+      # attribute as described in the +link_to+ documentation.
```
This links back to the documentation I added for [`data-turbo-method`](https://github.com/rails/rails/pull/47505/files#diff-68404eacab886d3ef2d43242eb40e7e66d4ff6135433ecc8a821faffc6104586R177) in the `link_to` docs. We did the same previously for the `:method` argument so I thought of doing the same for `data-turbo-method`. However, I'm happy to remove it if we feel it's not necessary as well!

julianfssen commented:
Thank you for the re-review, @zzak ! I appreciate your help. I've updated the PR with your feedback, let me know what you think!

julianfssen commented:
@zzak @ghiculescu , I'm bumping this PR again to decide if we should mention `data-turbo-method` and `data-turbo-confirm` in the `link_to` docs. 

I appreciate your previous feedback on this. Do you mind giving it another review to see if it's worth an addition to the docs?

I apologize for the delay and I understand if we feel we don't need to add this in the docs. Thank you!

ghiculescu commented on actionview/lib/action_view/helpers/url_helper.rb:
```diff
@@ -173,6 +173,34 @@ def _filtered_referrer # :nodoc:
+      # {Consult the Turbo Handbook for more information on performing visits with different
+      # methods.}[https://turbo.hotwired.dev/handbook/drive#performing-visits-with-a-different-method]
+      #
+      # A common use case is to insert links to perform update or delete operations
+      # (i.e. non-<tt>GET</tt> requests) in a form.
+      #
+      # For example, you might use +button_to+ in an edit form to delete a record:
+      #
+      #   button_to "Delete profile", @profile, method: :delete
+      #
+      # Since forms cannot appear inside another form,
+      # +button_to+ will not work here as it returns a form.
```
Is still find this whole section confusing. The bit about putting links inside forms being the most common usage doesn't sound right to me - is that actually the primary use case for this feature?

julianfssen replied:
@ghiculescu , you're right that it's likely not common to put links in forms. 

I've updated the docs to be much more concise when documenting the `data-turbo-method` and `data-turbo-confirm` options. What do you think of this change?

https://github.com/rails/rails/blob/6a798d079a875a5e60ace20ead23bb3e89eec712/actionview/lib/action_view/helpers/url_helper.rb#L176-L197
